x-pack/plugins/inference/server/chat_complete/adapters/gemini/gemini_adapter.ts

```diff
@@ -0,0 +1,222 @@
+        subAction: 'invokeStream',
+        subActionParams: {
+          messages: messagesToGemini({ system, messages, toolChoice }),
+          tools: toolsToGemini(tools),
+          temperature: 0,
+          stopSequences: ['\n\nHuman:'],
+        },
+      })
+    ).pipe(
+      switchMap((response) => {
+        const readable = response.data as Readable;
+        return eventSourceStreamIntoObservable(readable);
+      }),
+      map((line) => {
+        return JSON.parse(line) as GenerateContentResponseChunk;
+      }),
+      processVertexStream()
+    );
+  },
+};
+
+function toolsToGemini(tools: ToolOptions['tools']): Gemini.Tool[] {
+  return [
+    {
+      functionDeclarations: Object.entries(tools ?? {}).map(
+        ([toolName, { description, schema }]) => {
+          return {
+            name: toolName,
+            description,
+            parameters: schema
+              ? toolSchemaToGemini({ schema })
+              : {
+                  type: Gemini.FunctionDeclarationSchemaType.OBJECT,
+                  properties: {},
+                },
+          };
+        }
+      ),
+    },
+  ];
+}
+
+function toolSchemaToGemini({ schema }: { schema: ToolSchema }): Gemini.FunctionDeclarationSchema {
+  const convertSchemaType = ({
+    def,
+  }: {
+    def: ToolSchemaType;
+  }): Gemini.FunctionDeclarationSchemaProperty => {
+    switch (def.type) {
+      case 'array':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.ARRAY,
+          description: def.description,
+          items: convertSchemaType({ def: def.items }) as Gemini.FunctionDeclarationSchema,
+        };
+      case 'object':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.OBJECT,
+          description: def.description,
+          required: def.required as string[],
+          properties: Object.entries(def.properties).reduce<
+            Record<string, Gemini.FunctionDeclarationSchema>
+          >((properties, [key, prop]) => {
+            properties[key] = convertSchemaType({ def: prop }) as Gemini.FunctionDeclarationSchema;
+            return properties;
+          }, {}),
+        };
+      case 'string':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.STRING,
+          description: def.description,
+          enum: def.enum ? (def.enum as string[]) : def.const ? [def.const] : undefined,
+        };
+      case 'boolean':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.BOOLEAN,
+          description: def.description,
+          enum: def.enum ? (def.enum as string[]) : def.const ? [def.const] : undefined,
+        };
+      case 'number':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.NUMBER,
+          description: def.description,
+          enum: def.enum ? (def.enum as string[]) : def.const ? [def.const] : undefined,
+        };
+      default:
+        throw new Error(`Error converting Invalid parameter type: ${def}`);
```

Comment:
is there a way we can have this be a type error instead of creating a runtime error?

Reply:
Removing the `default` should do just that (`def` is `never` here, just kept the statement because of old habits)